### ISSUE TITLE
Enhancement to the Client Resource Theme Objects for Including or Excluding CSS and JS resources

### ISSUE THREAD

**Issue by ajplopez** [dnnsoftware/Dnn.Platform#4385]
## Description of problem
As a theme developer, I want to remove default.css from loading for unauthenticated users to improve the load time and avoid conflicts/overrides in the default.css file. I know I can use this in my .ascx skin: <dnn:DnnCssExclude runat=“server” Name=“dnndefault” />. However, this unloads the entire file for all users, including SuperUsers/Admins/Content Editors and other roles that depend on CSS inside that file. 

## Description of solution

If there was an attribute in the control, <dnn:DnnCssExclude runat=“server” Name=“dnndefault” />, where we could define which roles should be affected, then we can conditionally load the file easily. For example, perhaps we could define Roles="Unauthenticated" so it would only apply for unauthenticated users. Like this:

<dnn:DnnCssExclude runat=“server” Name=“dnndefault” Roles=“Unauthenticated” />  

The above would unload for only unauthenticated users. Or you could do something like this:

<dnn:DnnCssExclude runat=“server” Name=“dnndefault” Roles=“Unauthenticated,Content Editors” />  

The above would apply to both Unauthenticated and Content Editors.

## Description of alternatives considered
Alternatively, but also additionally, it would be useful if this Roles="SuperUsers" could be applied to the DnnCssInclude and DnnJsInclude like this:

<dnn:DnnCssInclude runat="server" FilePath="css/content-editor.css" PathNameAlias="SkinPath" Roles="SuperUsers,Administrators,Content Editors" />

With this method, we can add in some conditional CSS for different roles, like a Content Editor role. 

We could even have things like Roles="!Content Editors" to load for everyone besides Content Editors.

## Additional context
I'm trying to improve the front end of the websites I build but also avoid default.css typography issues. I'm using Tailwind to build sites and I don't really want to simply override all the time.

40Finger Style Helper Object already does this (and I've used it for years) but I don't want to install yet another skin object/module, and I don't want to depend on it. It would be better if these things came bundled with DNN already.

https://www.40fingers.net/Products/DNN-Stylehelper

## Affected browser
* all

**Comment by mitchelsellers** [comment]
You can accomplish this today with a code block in your theme.

Not saying it isn't worth investigating adding, but giving options

**Comment by ajplopez** [comment]
I've experimented with methods over the last couple of years, but using DnnCssExclude within a code block in my .ascx for SuperUsers or Admins only still runs and so the default.css is removed for them too. Not sure if you have a different block of code?

**Comment by jeremy-farrance** [comment]
I would like to recommend the title of this issue be changed, as it is kind of misleading. 
 
Suggestion? Maybe, "**Enhancement to the Client Resource Theme Objects for Including or Excluding CSS and JS resources**"

**Comment by parasdaryanani** [comment]
I usually do something like this in my skins (see below). This snippet checks the current user's role and removes the dnndefault stylesheet. This code is verbatim copied from the DnnCssExclude control with an added conditional statement around it that checks the user role.

```csharp
<script runat="server">
  protected override void OnPreRender(EventArgs e) {
    base.OnPreRender(e);
  
    var userInfo = UserController.GetCurrentUserInfo();
    if (!userInfo.IsInRole("Administrators")) {
      var loader = this.Page.FindControl("ClientResourceIncludes");
      if (loader != null) {
        ClientDependency.Core.Controls.ClientDependencyInclude ctlToRemove = null;
        foreach(ClientDependency.Core.Controls.ClientDependencyInclude ctl in loader.Controls) {
          if (ctl.Name.ToLowerInvariant() == "dnndefault" && ctl.DependencyType == ClientDependency.Core.ClientDependencyType.Css) {
            ctlToRemove = ctl;
            break;
          }
        }
  
        if (ctlToRemove != null) {
          loader.Controls.Remove(ctlToRemove);
        }
      }
    }
  }
</script>
```
If this is something the community decides to include into the DNN Core, it's a quick addition as above, assuming I haven't missed anything. Otherwise, this is a tried and tested workaround, without 3rd party extensions.

**Comment by ajplopez** [comment]
Thanks for this, Paras. Definitely something to keep in mind. 

Ultimately I went with another route using the DNNCSSExclude tag but also using 2sxc to conditonally link the default.css back in. But I also wrote overriding css in the skin package to handle the typography when default.css is loaded. I wrote a blog post about my method here: https://wolfxmachina.com/blog/category/dnn/unload-and-conditionally-load-defaultcss-in-dnn
